### PR TITLE
Add light effects

### PR DESCRIPTION
**Pull request by 857b**
Create a new shader "LightSpriteShader" which is based on SpriteShader. It can take as parameters the positions and the light emissions of the system stars and the ambient light of the system.
The objects far from the stars will be displayed as if they receive less light.
The shader also compute the normal for each pixel to create a shadow on the part which aren't facing the light sources. It assume that the object is spherical and is now only activated for the StellarObjects.
light effect
If the shader is not supported by the player hardware, the classic SpriteShader is used instead.

### CODE REVIEW

File: source/DrawList.cpp [hunk 1]
```diff
@@ -18,17 +18,32 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 #include "Sprite.h"
 #include "SpriteSet.h"
 #include "SpriteShader.h"
+#include "LightSpriteShader.h"
```
Comment (tehhowch) [inline source/DrawList.cpp]: See the C++ Style guide on the wiki (includes should be alphabetized)

File: source/DrawList.cpp [hunk 2]
```diff
@@ -43,85 +58,93 @@ void DrawList::SetCenter(const Point &center, const Point &centerVelocity)
 }
 
 
+void DrawList::setAmbiant(const float ambiant[3]){
```
Comment (tehhowch) [inline source/DrawList.cpp]: functions are CaptializedCamelCase

**Comment by Bladewood** [comment]
Oh, God...

File: source/DrawList.cpp [hunk 3]
```diff
@@ -43,85 +58,93 @@ void DrawList::SetCenter(const Point &center, const Point &centerVelocity)
...
-	Push(body, position, blur, 0., 1., swizzle);
-	return true;
+bool DrawList::AddStar(const Body& body, bool blur){	
+	return CullPush(body, body.Position(), blur ? body.Velocity() : centerVelocity, 0., 1., body.GetSwizzle(),false);
 }
 
 
+bool DrawList::AddLightSource(const float pos[3], const float emit[3]){
+	if(lightsPos.size()/3 >= (unsigned int)(LightSpriteShader::MaxNbLight)) return false;
+	for(int i=0;i<3;i++){
```
Comment (tehhowch) [inline source/DrawList.cpp]: Spacings between arguments

File: source/DrawList.cpp [hunk 4]
```diff
@@ -43,85 +58,93 @@ void DrawList::SetCenter(const Point &center, const Point &centerVelocity)
...
+	if(!lights || !LightSpriteShader::IsAvailable()){
+		SpriteShader::Bind();
+
+		for(const Item &item : items)
+			SpriteShader::Add(
+				item.tex0, item.tex1, item.position, item.transform,
+				item.Swizzle(), item.Clip(), item.Fade(), showBlur ? item.blur : nullptr);
+
+		SpriteShader::Unbind();
+	}else{
```
Comment (tehhowch) [inline source/DrawList.cpp]: Separate lines

File: source/DrawList.cpp [hunk 5]
```diff
@@ -43,85 +58,93 @@ void DrawList::SetCenter(const Point &center, const Point &centerVelocity)
...
+		LightSpriteShader::Bind();
+
+		for(const Item &item : items)
+			LightSpriteShader::Add(
+				item.tex0, item.tex1, item.position, item.transform,
+				item.Swizzle(), item.Clip(), item.Fade(), showBlur ? item.blur : nullptr,
+				item.posGS, item.transformGS, 
+				(item.Lighted()&&lights)?(lightsPos.size()/3):-1,lightAmbiant,lightsPos.data(),lightsEmit.data(), item.NormalUse());
```
Comment (tehhowch) [inline source/DrawList.cpp]: Spaces

File: source/DrawList.h [hunk 6]
```diff
@@ -30,35 +30,55 @@ class Sprite;
...
 	
-	void Push(const Body &body, Point pos, Point blur, double cloak, double clip, int swizzle);
+	void Push(const Body &body, Point posRel, Point posGS, Point blurRel, double cloak, double clip, int swizzle, bool lighted = false, float normUse = 0.f);
+	bool CullPush(const Body &body, Point posGS, Point vel, double cloak, double clip, int swizzle, bool lighted = false, float normUse = 0.f);
 	
 	
 private:
 	class Item {
 	public:
+		//constants for the flags
```
Comment (tehhowch) [inline source/DrawList.h]: Constants should be ALL_CAPS_UNDERSCORED

File: source/LightSpriteShader.cpp [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+#include "LightSpriteShader.h"
+
+#include "Point.h"
+#include "Screen.h"
+#include "Shader.h"
+#include "Sprite.h"
+
+#include <vector>
+#include <stdexcept>
```
Comment (tehhowch) [inline source/LightSpriteShader.cpp]: s < v

File: source/LightSpriteShader.h [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#ifndef LIGHT_SPRITE_SHADER_H_
+#define LIGHT_SPRITE_SHADER_H_
+
+class Sprite;
+class Point;
```
Comment (tehhowch) [inline source/LightSpriteShader.h]: P < S

File: source/DrawList.cpp (hunk 2 above)
Comment (Bladewood) [inline source/DrawList.cpp]: also, ambient*?

**Comment by tehhowch** [comment]
Regarding the shadow drawing: It's my understanding that the stellar object sprites all have a common base orientation, and this sprite is rotated based on the position of the stellar object in the system, relative to the star - e.g. base lighting+shadows are already "drawn" for stellar objects.

Also, does this scale the brightness effect based on the number of stars in the system, and the type / color of star? I have an existing PR to add star count information (so you could scale the brightness per system), but haven't looked into anything regarding the star color (presumably based on the filename of the star sprite.

**Comment by tehhowch** [comment]
You've also made an error in PRing from your `master` branch.

Your `master` branch should always be a 1:1 copy of the official upstream repo branch, so that you can create new features based on a current version of the game without affecting your existing pull request.
If you review the other pulls, you'll note that we create a feature branch entirely devoted to the topic at hand, and make the PR from that, and only add commits required for that feature to that branch.

I recommend closing this PR, and creating a new branch based on your current master:
```
git checkout -b SolarDistanceLighting
git push -u origin SolarDistanceLighting
```
then restore your master to a 1:1 copy of the official repo
```
git checkout master
git fetch upstream
git reset --hard upstream/master
git push -f origin
```
Then you can open the new PR from your feature branch.

**Pull request closed without merging**